File: apps/flutter_sample_spm/ios/Runner/PermissionChannelHandler.swift
import Flutter
import CoreLocation
import UIKit

class PermissionChannelHandler: NSObject, CLLocationManagerDelegate {
    private let locationManager = CLLocationManager()
    // Pending result for the foreground (When In Use) request, resolved by the delegate.
    private var locationPermissionResult: FlutterResult?

    func register(with messenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(
            name: "io.customer.testbed/permissions",
            binaryMessenger: messenger
        )
        channel.setMethodCallHandler(handle)
        locationManager.delegate = self
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "requestNotificationPermission":
            requestNotificationPermission(result: result)
        case "getNotificationPermissionStatus":
            getNotificationPermissionStatus(result: result)
        case "requestLocationPermission":
            requestLocationPermission(result: result)
        case "requestBackgroundLocationPermission":
            requestBackgroundLocationPermission(result: result)
        case "getLocationAuthorizationStatus":
            getLocationAuthorizationStatus(result: result)
        case "openAppSettings":
            if let url = URL(string: UIApplication.openSettingsURLString) {
                UIApplication.shared.open(url)
            }
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // Drives the state-aware "Background Location" button, mirroring the native sample apps.
    private func getLocationAuthorizationStatus(result: @escaping FlutterResult) {
        switch locationManager.authorizationStatus {
        case .authorizedAlways:
            result("backgroundGranted")
        case .authorizedWhenInUse:
            result("foregroundOnly")
        case .denied, .restricted:
            result("denied")
        case .notDetermined:
            result("notDetermined")
        @unknown default:
            result("denied")
        }
    }

    private func getNotificationPermissionStatus(result: @escaping FlutterResult) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            DispatchQueue.main.async {
                switch settings.authorizationStatus {
                case .authorized, .provisional, .ephemeral:
                    result("granted")
                case .denied:
                    result("permanentlyDenied")
                case .notDetermined:
                    result("notDetermined")
                @unknown default:
                    result("denied")
                }
            }
        }
    }

    private func requestNotificationPermission(result: @escaping FlutterResult) {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, _ in
            DispatchQueue.main.async {
                result(granted ? "granted" : "denied")
            }
        }
    }

    // Only one permission request can be in flight at a time (single result slot).
    // Reject a second concurrent request instead of clobbering the first's callback.
    private func claimPendingResult(_ result: @escaping FlutterResult) -> Bool {
        if locationPermissionResult != nil {
            result("denied")
            return false
        }
        locationPermissionResult = result
        return true
    }

    private func requestLocationPermission(result: @escaping FlutterResult) {
        let status = locationManager.authorizationStatus
        switch status {
        case .authorizedWhenInUse, .authorizedAlways:
            result("granted")
        case .denied, .restricted:
            result("permanentlyDenied")
        case .notDetermined:
            guard claimPendingResult(result) else { return }
            locationManager.requestWhenInUseAuthorization()
        @unknown default:
            result("denied")
        }
    }

    // "Always" authorization is required for geofence region monitoring to wake the app
    // in the background. iOS escalates from When In Use to Always with a separate prompt.
    private func requestBackgroundLocationPermission(result: @escaping FlutterResult) {
        let status = locationManager.authorizationStatus
        switch status {
        case .authorizedAlways:
            result("granted")
        case .denied, .restricted:
            result("permanentlyDenied")
        case .notDetermined, .authorizedWhenInUse:
            // The Always upgrade is delivered asynchronously, and Core Location may never
            // call the delegate (e.g. the user keeps "While Using", or the prompt was
            // already shown once). So don't await it — kick off the request and report
            // "pending"; the UI reflects the real outcome on resume via the status query.
            locationManager.requestAlwaysAuthorization()
            result("pending")
        @unknown default:
            result("denied")
        }
    }

    // Resolves the foreground (When In Use) request. The background request is fire-and-forget,
    // so When In Use here is always a foreground grant.
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        guard let result = locationPermissionResult else { return }
        locationPermissionResult = nil

        switch manager.authorizationStatus {
        case .authorizedWhenInUse, .authorizedAlways:
            result("granted")
        case .denied, .restricted:
            result("permanentlyDenied")
        default:
            result("denied")
        }
    }
}

File: apps/shared/android/io/customer/testbed/flutter/PermissionChannelHandler.kt
package io.customer.testbed.flutter

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class PermissionChannelHandler(
    private val activity: Activity
) : MethodChannel.MethodCallHandler {

    companion object {
        private const val REQUEST_NOTIFICATION = 1001
        private const val REQUEST_LOCATION = 1002
        private const val REQUEST_BACKGROUND_LOCATION = 1003
    }

    private var pendingResult: MethodChannel.Result? = null

    fun register(channel: MethodChannel) {
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "requestNotificationPermission" -> requestNotificationPermission(result)
            "getNotificationPermissionStatus" -> getNotificationPermissionStatus(result)
            "requestLocationPermission" -> requestLocationPermission(result)
            "requestBackgroundLocationPermission" -> requestBackgroundLocationPermission(result)
            "getLocationAuthorizationStatus" -> getLocationAuthorizationStatus(result)
            "openAppSettings" -> {
                val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                    data = Uri.fromParts("package", activity.packageName, null)
                }
                activity.startActivity(intent)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // Drives the state-aware "Background Location" button, mirroring the native sample apps.
    private fun getLocationAuthorizationStatus(result: MethodChannel.Result) {
        val fineGranted = ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION) ==
            PackageManager.PERMISSION_GRANTED
        if (!fineGranted) {
            val permanentlyDenied = !ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.ACCESS_FINE_LOCATION) &&
                hasRequestedPermissionBefore(Manifest.permission.ACCESS_FINE_LOCATION)
            result.success(if (permanentlyDenied) "denied" else "notDetermined")
            return
        }
        val backgroundGranted = Build.VERSION.SDK_INT < Build.VERSION_CODES.Q ||
            ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_BACKGROUND_LOCATION) ==
            PackageManager.PERMISSION_GRANTED
        result.success(if (backgroundGranted) "backgroundGranted" else "foregroundOnly")
    }

    private fun getNotificationPermissionStatus(result: MethodChannel.Result) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            result.success("granted")
            return
        }
        when {
            ContextCompat.checkSelfPermission(activity, Manifest.permission.POST_NOTIFICATIONS)
                    == PackageManager.PERMISSION_GRANTED -> result.success("granted")
            ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.POST_NOTIFICATIONS)
                    -> result.success("denied")
            else -> result.success("notDetermined")
        }
    }

    private fun requestNotificationPermission(result: MethodChannel.Result) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            result.success("granted")
            return
        }
        if (ContextCompat.checkSelfPermission(activity, Manifest.permission.POST_NOTIFICATIONS)
            == PackageManager.PERMISSION_GRANTED) {
            result.success("granted")
            return
        }
        if (!claimPendingResult(result)) return
        ActivityCompat.requestPermissions(
            activity,
            arrayOf(Manifest.permission.POST_NOTIFICATIONS),
            REQUEST_NOTIFICATION
        )
    }

    private fun requestLocationPermission(result: MethodChannel.Result) {
        if (ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION)
            == PackageManager.PERMISSION_GRANTED) {
            result.success("granted")
            return
        }
        if (!ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.ACCESS_FINE_LOCATION)
            && hasRequestedPermissionBefore(Manifest.permission.ACCESS_FINE_LOCATION)) {
            result.success("permanentlyDenied")
            return
        }
        if (!claimPendingResult(result)) return
        markPermissionRequested(Manifest.permission.ACCESS_FINE_LOCATION)
        ActivityCompat.requestPermissions(
            activity,
            arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
            REQUEST_LOCATION
        )
    }

    // Background location is a separate, escalated grant required for geofence transition
    // delivery while the app is backgrounded. It only exists as its own permission on
    // API 29+, and on API 30+ the OS requires fine location to be granted first.
    private fun requestBackgroundLocationPermission(result: MethodChannel.Result) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            result.success("granted")
            return
        }
        if (ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION)
            != PackageManager.PERMISSION_GRANTED) {
            result.success("fineLocationRequired")
            return
        }
        if (ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_BACKGROUND_LOCATION)
            == PackageManager.PERMISSION_GRANTED) {
            result.success("granted")
            return
        }
        if (!ActivityCompat.shouldShowRequestPermissionRationale(activity, Manifest.permission.ACCESS_BACKGROUND_LOCATION)
            && hasRequestedPermissionBefore(Manifest.permission.ACCESS_BACKGROUND_LOCATION)) {
            result.success("permanentlyDenied")
            return
        }
        if (!claimPendingResult(result)) return
        markPermissionRequested(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
        ActivityCompat.requestPermissions(
            activity,
            arrayOf(Manifest.permission.ACCESS_BACKGROUND_LOCATION),
            REQUEST_BACKGROUND_LOCATION
        )
    }

    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray): Boolean {
        if (requestCode != REQUEST_NOTIFICATION &&
            requestCode != REQUEST_LOCATION &&
            requestCode != REQUEST_BACKGROUND_LOCATION
        ) return false

        val result = pendingResult ?: return false
        pendingResult = null

        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            result.success("granted")
        } else {
            val permission = permissions.firstOrNull() ?: ""
            if (!ActivityCompat.shouldShowRequestPermissionRationale(activity, permission)) {
                result.success("permanentlyDenied")
            } else {
                result.success("denied")
            }
        }
        return true
    }

    // Only one permission request can be in flight at a time (single result slot).
    // Reject a second concurrent request instead of clobbering the first's callback.
    private fun claimPendingResult(result: MethodChannel.Result): Boolean {
        if (pendingResult != null) {
            result.success("denied")
            return false
        }
        pendingResult = result
        return true
    }

    private fun hasRequestedPermissionBefore(permission: String): Boolean {
        return activity.getSharedPreferences("permission_prefs", Activity.MODE_PRIVATE)
            .getBoolean(permission, false)
    }

    private fun markPermissionRequested(permission: String) {
        activity.getSharedPreferences("permission_prefs", Activity.MODE_PRIVATE)
            .edit().putBoolean(permission, true).apply()
    }
}
